plugin.video.mediathekview/resources/lib/downloader.py

```diff
@@ -0,0 +1,361 @@
+# -*- coding: utf-8 -*-
+"""
+Download handler module
+
+Copyright (c) 2018, Leo Moll
+Licensed under MIT License
+"""
+
+# pylint: disable=import-error
+# pylint: disable=mixed-indentation, bad-whitespace, bad-continuation, missing-docstring
+
+# -- Imports ------------------------------------------------
+from __future__ import unicode_literals
+
+import os
+import re
+
+from contextlib import closing
+
+import xbmc
+import xbmcvfs
+
+import resources.lib.mvutils as mvutils
+
+from resources.lib.kodi.KodiUI import KodiBGDialog
+from resources.lib.filmui import FilmUI
+from resources.lib.ttml2srt import ttml2srt
+
+# -- Classes ------------------------------------------------
+class Downloader( object ):
+	def __init__( self, plugin ):
+		self.plugin		= plugin
+		self.database	= plugin.database
+		self.settings	= plugin.settings
+		self.notifier	= plugin.notifier
+
+	def play_movie_with_subs( self, filmid, external ):
+		film = self.database.RetrieveFilmInfo( filmid )
+		if film is None:
+			self.notifier.ShowError( 30990, self.plugin.language( 30991 ) )
+			return
+		ttmname = self.plugin.datapath + '/subtitle.ttml'
+		srtname = self.plugin.datapath + '/subtitle.srt'
+		subs = []
+		if self.download_subtitle( film, ttmname, srtname, 'subtitle' ):
+			subs.append( srtname )
+		( videourl, listitem ) = FilmUI( self.plugin ).get_list_item( None, film )
+		if listitem:
+			if subs:
+				listitem.setSubtitles( subs )
+			if external:
+				xbmc.Player().play( videourl, listitem )
```

Comment:
why you need to call `xbmc.Player()`? what is external exactly? An external player?
Note that calling xbmc.Player in plugins is a bad practice, an entry in the db will be created each time a user plays a file. Plugin:// paths, on the other hand, are unique.

Reply:
The routine `def play_movie_with_subs( self, filmid, external ):` has a double purpose. But first let me explain the circumstances:

The german public broadcasters have a URL with a (playable) media content (usually MP4). In addition many shows have also an additional URL for subtitles. Unfortunately these subtitles are not playable, since they are provided in an _"unusual"_ format that - as far as I know - is not supported by Kodi (TTML and in rare cases VTT).

If you want to play such a stream with subtitles, you have to download the TTML subtitle **BEFORE** playing the stream, convert it, create a new list item and **THAN** play the stream.

Regarding playing shows with subtitles there are two operating modes in the plugin, in dependency of a configuration setting.

If "Autoplay Subtitles" is on, the following happens: when the directory list is filled, each movie with subtitles will not be added with his real media url, but with the plugin URL itself and as parameters `{'mode': "playwithsrt", 'id': <MOVIEID>, 'external': False}`. This is a kind of deferred media URL retrieval mechanism ([See this forum post](https://forum.kodi.tv/showthread.php?tid=173986)). When the user selects the entry in the directory for playing, Kodi notices that the media URL is not a playable video URL and than invokes the specified plugin URL expecting the real URL to be resolved. Here the Plugin passes the supplied parameters to `play_movie_with_subs(<MOVIEID>, False)`. The method now downloads the incompatible subtitle file locally, converts it to a supported format fills in the information into a `listitem` and returns the information to Kodi by invoking `xbmcplugin.setResolvedUrl( True, listitem )` ([See Documentation](https://codedocs.xyz/AlwinEsch/kodi/group__python__xbmcplugin.html#gaa03ad528583db7d5277eae3977acde0f)). Kodi starts to play the media with the subtitles.

If "Autoplay Subtitles" is off, the following happens: when the directory list is filled, each movie will be added with his real media URL but with an additional context menu entry (_"Play with Subtitles"_). When the user specifically selects the context menu item _"Play with Subtitles"_, the Plugin itself will be invoked with the parameters `{'mode': "playwithsrt", 'id': <MOVIEID>, 'external': True}`. Here the Plugin passes the supplied parameters to `play_movie_with_subs(<MOVIEID>, True)`. The method now downloads the incompatible subtitle file locally, converts it to a supported format fills in the information into a `listitem` and than invokes `xbmc.Player().play( videourl, listitem )` because this is the only way known to me for playing an item from a command posted into the plugin.

If you have any better idea, please let me know. I invested quite a lot of time in finding this solution but I will be very glad to implement anything better, if possible.

Reply:
Yeah Kodi lacks support for ttml based subtitles (although inputstream.adaptive already has this functionality for mpd and hls manifests). Definitely something to look for in the future.
The explanation looks good to me and I can see your efforts in doing the logic properly. I am wondering, in the second case where you call `xbmc.Player().play()`, why not to simply use a builtin-function with the `plugin://` path instead of this class? 
Example:
`xbmc.executebuiltin("PlayMedia(plugin://yourplugin/arg/)")`

Another option could be to use the JSON-RPC (xbmc.executeJSONRPC with a `Player.Open` action). Would this work for you?

Reply:
Thank you for the suggestions! The `xbmc.executebuiltin("PlayMedia(plugin://yourplugin/arg/)")` sounds very promising. I will give it a try and if it works, I will implement it in one of the next versions.

Reply:
Tracked Issue in https://github.com/mediathekview/plugin.video.mediathekview/issues/88